File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/handler/check/impl/DoubleValidator.java
package io.github.oleksivio.telegram.bot.core.controller.handler.check.impl;

import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator;

public class DoubleValidator implements Validator<Double> {

    private final double lessThan;
    private final double lessOrEqual;
    private final double moreThan;
    private final double moreOrEqual;
    private final double equal;

    public DoubleValidator(double lessThan,
                           double lessOrEqual,
                           double moreThan,
                           double moreOrEqual,
                           double equal) {
        this.lessThan = lessThan;
        this.lessOrEqual = lessOrEqual;
        this.moreThan = moreThan;
        this.moreOrEqual = moreOrEqual;
        this.equal = equal;
    }

    @Override
    public boolean check(Double argument) {
        if (!Double.isNaN(equal) && Double.compare(argument, equal) != 0) {
            return false;
        }

        return argument < lessThan
                && argument <= lessOrEqual
                && argument > moreThan
                && argument >= moreOrEqual;
    }
}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/handler/check/impl/DoubleValidator.kt
package io.github.oleksivio.telegram.bot.core.controller.handler.check.impl

import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator

class DoubleValidator(private val lessThan: Double,
                      private val lessOrEqual: Double,
                      private val moreThan: Double,
                      private val moreOrEqual: Double,
                      private val equal: Double) : Validator<Double> {
    override fun invoke(checkedValue: Double): Boolean {
        return if (!java.lang.Double.isNaN(equal) && java.lang.Double.compare(checkedValue, equal) != 0) {
            false
        } else checkedValue < lessThan
                && checkedValue <= lessOrEqual
                && checkedValue > moreThan
                && checkedValue >= moreOrEqual
    }


}
